### PR TITLE
Update type inference doc about regions

### PR DESCRIPTION
**Pull request by zhouxin913**
Fix https://github.com/openxla/stablehlo/issues/762

1. Clarify the updated split rules among verifiers and shape function
2. Clarify how to write / maintain the tests.

Also add @sdasgup3 as CC who is in interested.

### CODE REVIEW

File: docs/type_inference.md
Comment (burmako) [inline docs/type_inference.md:31]: "the autogenerated `build()` functions"

File: docs/type_inference.md
Comment (burmako) [inline docs/type_inference.md:37]: Let's write "\`PadOp\`" for consistency with the second bullet.

Comment (burmako) [inline docs/type_inference.md:37]: "without regions". Let's expand the abbreviation here since it doesn't save that much typing.

File: docs/type_inference.md
Comment (burmako) [inline docs/type_inference.md:40]: "with regions".

Comment (burmako) [inline docs/type_inference.md:40]: "the autogenerated builders don't take regions as parameters, so if these builders involve type inference, then the shape function will be called with empty regions".

Comment (burmako) [inline docs/type_inference.md:40]: Use a permalink here: https://github.com/tensorflow/mlir-hlo/blob/129ae36971a9e3e110d8b91b91a150942d13ff81/mhlo/transforms/mhlo_canonicalize_reduction/mhlo_canonicalize_reduction.cc#L221.

File: docs/type_inference.md
Comment (burmako) [inline docs/type_inference.md:42]: "If the regions are needed for type inference (`IfOp/CaseOp/MapOp`), then additionally the shape function...".

File: docs/type_inference.md
Comment (burmako) [inline docs/type_inference.md:59]: Let's add a newline between the header and the text.

File: docs/type_inference.md
Comment (burmako) [inline docs/type_inference.md:60]: "When implementing or revisiting...".

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/type_inference.md
Comment (smit-hinsu) [inline docs/type_inference.md:41]: why not suggest use of verifyRegions so that verification that is part of type inference doesn't need to happen again? by the time verifyRegions is invoked, type inference would have been done already. This should help avoid most of the duplication.

File: docs/type_inference.md
Comment (zhouxin913) [reply docs/type_inference.md:40]: Thanks!

File: docs/type_inference.md
Comment (zhouxin913) [reply docs/type_inference.md:41]: Hi Smit, thanks for the suggestions: `verifyRegions()` seems not used in StableHLO / MHLO yet, but it's a good option. For the `ReduceOp`, we could consider put the region-related verification (link below)  into a new `verifyRegions()` method, but as you may notice, it relies on a lot of intermediate result from other parts of code above it. It it not clear enough for me for the moment to split them further.

As the crash that https://github.com/openxla/stablehlo/pull/760 and this PR trying to address is an P0 issue, I will submit after get approval. But we may continue the discussion in a new bug if necessary. 

region-related verification in `ReduceOp`: https://github.com/tensorflow/mlir-hlo/blame/master/stablehlo/stablehlo/dialect/TypeInference.cpp#L1551

Comment (zhouxin913) [reply docs/type_inference.md]: Mark as resolved for now, and still open to discussion in the future.

**Review by burmako — APPROVED** [review]

**Pull request merged**
